File: core/src/main/java/com/graphhopper/routing/util/parsers/OSMHazmatTunnelParser.java
package com.graphhopper.routing.util.parsers;

import com.graphhopper.reader.ReaderWay;
import com.graphhopper.routing.ev.EdgeBytesAccess;
import com.graphhopper.routing.ev.EnumEncodedValue;
import com.graphhopper.routing.ev.HazmatTunnel;
import com.graphhopper.storage.BytesRef;

public class OSMHazmatTunnelParser implements TagParser {

    private static final String[] TUNNEL_CATEGORY_NAMES;

    static {
        HazmatTunnel[] categories = HazmatTunnel.values();
        TUNNEL_CATEGORY_NAMES = new String[categories.length];
        for (int i = 0; i < categories.length; i++) {
            TUNNEL_CATEGORY_NAMES[i] = categories[i].name();
        }
    }

    private final EnumEncodedValue<HazmatTunnel> hazTunnelEnc;

    public OSMHazmatTunnelParser(EnumEncodedValue<HazmatTunnel> hazTunnelEnc) {
        this.hazTunnelEnc = hazTunnelEnc;
    }

    @Override
    public void handleWayTags(int edgeId, EdgeBytesAccess edgeAccess, ReaderWay readerWay, BytesRef relationFlags) {
        if (readerWay.hasTag("hazmat:adr_tunnel_cat", TUNNEL_CATEGORY_NAMES)) {
            HazmatTunnel code = HazmatTunnel.valueOf(readerWay.getTag("hazmat:adr_tunnel_cat"));
            hazTunnelEnc.setEnum(false, edgeId, edgeAccess, code);
        } else if (readerWay.hasTag("hazmat:tunnel_cat", TUNNEL_CATEGORY_NAMES)) {
            HazmatTunnel code = HazmatTunnel.valueOf(readerWay.getTag("hazmat:tunnel_cat"));
            hazTunnelEnc.setEnum(false, edgeId, edgeAccess, code);
        } else if (readerWay.hasTag("tunnel", "yes")) {
            HazmatTunnel[] codes = HazmatTunnel.values();
            for (int i = codes.length - 1; i >= 0; i--) {
                if (readerWay.hasTag("hazmat:" + codes[i].name(), "no")) {
                    hazTunnelEnc.setEnum(false, edgeId, edgeAccess, codes[i]);
                    break;
                }
            }
        }
    }
}

File: core/src/main/kotlin/com/graphhopper/routing/util/parsers/OSMHazmatTunnelParser.kt
package com.graphhopper.routing.util.parsers

import com.graphhopper.reader.ReaderWay
import com.graphhopper.routing.ev.EdgeIntAccess
import com.graphhopper.routing.ev.EnumEncodedValue
import com.graphhopper.routing.ev.HazmatTunnel
import com.graphhopper.storage.IntsRef

class OSMHazmatTunnelParser(private val hazTunnelEnc: EnumEncodedValue<HazmatTunnel>) : TagParser {

    override fun handleWayTags(edgeId: Int, edgeIntAccess: EdgeIntAccess, way: ReaderWay, relationFlags: IntsRef?) {
        if (way.hasTag("hazmat:adr_tunnel_cat", *TUNNEL_CATEGORY_NAMES)) {
            val code = HazmatTunnel.valueOf(way.getTag("hazmat:adr_tunnel_cat")!!)
            hazTunnelEnc.setEnum(false, edgeId, edgeIntAccess, code)
        } else if (way.hasTag("hazmat:tunnel_cat", *TUNNEL_CATEGORY_NAMES)) {
            val code = HazmatTunnel.valueOf(way.getTag("hazmat:tunnel_cat")!!)
            hazTunnelEnc.setEnum(false, edgeId, edgeIntAccess, code)
        } else if (way.hasTag("tunnel", "yes")) {
            val codes = HazmatTunnel.entries
            for (i in codes.indices.reversed()) {
                if (way.hasTag("hazmat:" + codes[i].name, "no")) {
                    hazTunnelEnc.setEnum(false, edgeId, edgeIntAccess, codes[i])
                    break
                }
            }
        }
    }

    companion object {
        private val TUNNEL_CATEGORY_NAMES: Array<String> =
            HazmatTunnel.entries.map { it.name }.toTypedArray()
    }
}
